### ISSUE TITLE
Custom Loss Function for Autoencoder

### ISSUE THREAD

**Issue by peoky** [keras-team/keras#10529]
Hi
I'm trying to build an auto-encoder in keras with a custom loss function, for example, consider the following auto-encoder:

```python
x = Input(shape=(50,))
encoded = Dense(32, activation='relu')(x)
decoded = Dense(50, activation='sigmoid')(encoded)

encoder = Model(x, encoded)
autoencoder = Model(x, decoded)

m1 = encoder.predict(c1) #c1, c2, c3: some inputs
m2 = encoder.predict(c2)
m3 = encoder.predict(c3)

autoencoder.compile(optimizer='adam', loss=custom_loss(encoded, m1, m2, m3))

from keras import backend as K
def custom_loss(x, m1, m2, m3):`
    def loss(y_true, y_pred):
        a = k.mean(k.square(y_true - y_pred))
        b = 0
        # assume our data belongs to 3 classes
        if x belongs to class1:
            b = k.mean(k.square(x - m1))
        elif x belongs to class2:
            b = k.mean(k.square(x - m2))
        elif x belongs to class3:
            b = k.mean(k.square(x - m3))
        return a + b
    return loss
```

I don't know how to implement custom loss to calculate 'b', also I need to update m1, m2, and m3 after every batch, can anyone plz help me?

**Comment by jemshit** [comment]
@peoky Hi, what was the problem with default loss and would you share your solution?

**Comment by peoky** [comment]
@jemshit : Hi, i did following solution:
```python
x = Input(shape=(50,))
m1 = Input(shape=(50,))
m2 = Input(shape=(50,))
m3 = Input(shape=(50,))

shared_dense = Dense(32, activation='relu')
encoded_x = shared_dense(x)
encoded_m1 = shared_dense(m1)
encoded_m2 = shared_dense(m2)
encoded_m3 = shared_dense(m3)

decoded = Dense(50, activation='sigmoid')(encoded_x)
predicted = Dense(3, activation='sigmoid')(decoded)

encoder = Model([x, m1, m2, m3], encoded_x)
autoencoder = Model([x, m1, m2, m3], [decoded, predicted])

autoencoder.compile(optimizer='adam', loss=custom_loss(encoded_x, encoded_m1, encoded_m2, encoded_m3))

def custom_loss(encoded_x, encoded_m1, encoded_m2, encoded_m3):
    def loss(y_true, y_pred):
        a = K.categorical_crossentropy(y_true, y_pred)
        a *= 0 # we don't need a, but without it, "An operation has `None` for gradient" error rises up
        b = 0
        for i in range(128):  # you have to set batch_size = 128
            b += K.switch(K.equal(K.argmax(y_true[i]), 0),
                    lambda: K.mean(K.square(encoded_x[i] - encoded_m1[i])),
                    lambda: K.switch(K.equal(K.argmax(y_true[i]), 1),
                                     lambda: K.mean(K.square(encoded_x[i] - encoded_m2[i])),
                                     lambda: K.mean(K.square(encoded_x[i] - encoded_m3[i]))))
        return a + b
    return loss

import numpy as np
x = np.random.rand(512, 50)
m1 = np.random.rand(512, 50)
m2 = np.random.rand(512, 50)
m3 = np.random.rand(512, 50)

y = np.random.randint(0, high=3, size=512)
y = keras.utils.to_categorical(y)

autoencoder.fit([x, m1, m2, m3], [x, y], batch_size=128, epochs=30)
